locator/azure_snitch.cc

```diff
@@ -0,0 +1,160 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/*
+ * Modified by ScyllaDB
+ * Copyright (C) 2021-present ScyllaDB
+ */
+
+#include "locator/azure_snitch.hh"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/seastar.hh>
+#include <seastar/http/response_parser.hh>
+#include <seastar/net/dns.hh>
+
+#include <boost/algorithm/string/classification.hpp>
+#include <boost/algorithm/string/split.hpp>
+#include <fmt/format.h>
+
+namespace locator {
+
+const auto azure_snitch::REGION_NAME_QUERY_PATH = fmt::format(AZURE_QUERY_PATH_TEMPLATE, "location");
+const auto azure_snitch::ZONE_NAME_QUERY_PATH = fmt::format(AZURE_QUERY_PATH_TEMPLATE, "zone");
+
+azure_snitch::azure_snitch(const sstring& fname, unsigned io_cpuid) : production_snitch_base(fname) {
+    if (this_shard_id() == io_cpuid) {
+        io_cpu_id() = io_cpuid;
+    }
+}
+
+future<> azure_snitch::load_config() {
+    sstring region = co_await azure_api_call(REGION_NAME_QUERY_PATH);
+    sstring azure_zone = co_await azure_api_call(ZONE_NAME_QUERY_PATH);
+
+    sstring datacenter_suffix = co_await read_property_file();
+
+    sstring azure_region = region + datacenter_suffix;
+
+    logger().info("AzureSnitch using region: {}, zone: {}.", azure_region, azure_zone);
+
+    _my_rack = azure_zone;
```

Comment:
problem is, that on some VMs I saw azure zone to be empty (while the VM had a zone)
so we have to be careful about defaults here

I didn't figure out on what does it depend, but fallback to getting the zone from somewhere (or defaulting to some at least empty string? or DEFAULT_RACK (I think we do it in some other snitch) should be good)

Reply:
OR the same goes for location
@fee-mendes do you still have some Azure VMs around?
can you do
```
curl -H Metadata:true --noproxy "*" "http://169.254.169.254/metadata/instance?api-version=2021-02-01&format=text"
```
(or the same with version 2018-04-02 )
on all your Azure VMs and share with @penberg ?